lib/internal/test_runner/snapshot.js

```diff
@@ -0,0 +1,234 @@
+  validateArray,
+  validateFunction,
+  validateObject,
+} = require('internal/validators');
+const { strictEqual } = require('assert');
+const { mkdirSync, readFileSync, writeFileSync } = require('fs');
+const { dirname } = require('path');
+const { createContext, runInContext } = require('vm');
+const kExperimentalWarning = 'Snapshot testing';
+const kMissingSnapshotTip = 'Missing snapshots can be generated by rerunning ' +
+  'the command with the --test-update-snapshots flag.';
+const defaultSerializers = [
+  (value) => { return JSONStringify(value, null, 2); },
+];
+
+function defaultResolveSnapshotPath(testPath) {
+  if (typeof testPath !== 'string') {
+    return testPath;
+  }
+
+  return `${testPath}.snapshot`;
+}
+
+let resolveSnapshotPathFn = defaultResolveSnapshotPath;
+let serializerFns = defaultSerializers;
+
+function setResolveSnapshotPath(fn) {
+  emitExperimentalWarning(kExperimentalWarning);
+  validateFunction(fn, 'fn');
+  resolveSnapshotPathFn = fn;
+}
+
+function setDefaultSerializers(serializers) {
+  emitExperimentalWarning(kExperimentalWarning);
+  validateFunctionArray(serializers, 'serializers');
+  serializerFns = ArrayPrototypeMap(serializers, (fn) => {
+    return fn;
+  });
+}
+
+class SnapshotManager {
+  constructor(entryFile, updateSnapshots) {
+    this.entryFile = entryFile;
+    this.snapshotFile = undefined;
+    this.snapshots = { __proto__: null };
+    this.nameCounts = new SafeMap();
+    // A manager instance will only read or write snapshot files based on the
+    // updateSnapshots argument.
+    this.loaded = updateSnapshots;
+    this.updateSnapshots = updateSnapshots;
+  }
+
+  resolveSnapshotFile() {
+    if (this.snapshotFile === undefined) {
+      const resolved = resolveSnapshotPathFn(this.entryFile);
+
+      if (typeof resolved !== 'string') {
+        const err = new ERR_INVALID_STATE('Invalid snapshot filename.');
+        err.filename = resolved;
+        throw err;
+      }
+
+      this.snapshotFile = resolved;
+    }
+  }
+
+  serialize(input, serializers = serializerFns) {
+    try {
+      let value = input;
+
+      for (let i = 0; i < serializers.length; ++i) {
+        const fn = serializers[i];
+        value = fn(value);
+      }
+
+      return `\n${templateEscape(value)}\n`;
+    } catch (err) {
+      const error = new ERR_INVALID_STATE(
+        'The provided serializers did not generate a string.',
+      );
+      error.input = input;
+      error.cause = err;
+      throw error;
+    }
+  }
+
+  getSnapshot(id) {
+    if (!(id in this.snapshots)) {
+      const err = new ERR_INVALID_STATE(`Snapshot '${id}' not found in ` +
+        `'${this.snapshotFile}.' ${kMissingSnapshotTip}`);
+      err.snapshot = id;
+      err.filename = this.snapshotFile;
+      throw err;
+    }
+
+    return this.snapshots[id];
+  }
+
+  setSnapshot(id, value) {
+    this.snapshots[templateEscape(id)] = value;
+  }
+
+  nextId(name) {
+    const count = this.nameCounts.get(name) ?? 1;
```

Comment:
There were concerns about this naming strategy in case a test is conditional (e.g. has a t.assert.snapsshot inside an `if`) last time

Reply:
AFAIK, that's just how snapshot testing works in general? The only way around that would be to parse the source code to identify all possible snapshot assertions.

Reply:
I assume we have the exact location of a test, so we can use that (line/column) as something more unique than the count?
this is something that other test runners are limited from implementing if I understand correctly

Reply:
The line changes very easily though. If you add a newline at the top of the file, it would unnecessarily invalidate all of the snapshots.

Reply:
fair point.

Reply:
maybe we should just document this caveat

Reply:
I don't think it's a caveat. I think it would render the whole API really annoying to use at all.

Reply:
I also think the fact that this API is scoped to just the test context now, instead of the more general assert module makes this much less of a concern. It also lines up with how the other test runners do it.